### ISSUE TITLE
XMB UI HiDPI (Wayland) scaling is inconsistent when a 3D core is running

### ISSUE THREAD

**Issue by myfreeweb** [libretro/RetroArch#6554]
## Description

I'm using a 4K screen with fractional HiDPI in Wayland (apps render at 2x, compostior downscales to 1.5x, I think macOS works similarly).

GLUI's scale is perfectly consistent: no matter what core I'm running. No core, 2D, 3D, it's always the same scale, *its* UI scale is all I need to set, Windowed Scale can be kept at 1, the text size is always correct.

**XMB**'s scale is all over the place, depending on the running core.

Setting Windowed Scale to 5 or **6** makes the UI size readable when no core is running, but cores start creating huge windows (I guess that's the actual intent of that setting), and UI size becomes even more ridiculous when a 3D core is running.

#### Windowed Scale = 1

mGBA:

![winscale1-mgba](https://user-images.githubusercontent.com/208340/38584818-38247b0e-3d20-11e8-8203-6678ea0e4c76.png)

Mupen64:

![winscale1-mupen](https://user-images.githubusercontent.com/208340/38584828-42d88f54-3d20-11e8-8492-3061fe13440f.png)

Reicast:

![winscale1-reicast](https://user-images.githubusercontent.com/208340/38584842-4c8e464c-3d20-11e8-9a14-a8d64c288119.png)

#### Windowed Scale = 2

No core:

![winscale2-blank](https://user-images.githubusercontent.com/208340/38584879-7867ba6e-3d20-11e8-877f-531fb6ee8933.png)

mGBA:

![winscale2-mgba](https://user-images.githubusercontent.com/208340/38584885-7f7c8dac-3d20-11e8-826f-c3141cdd7b97.png)

Mupen64:

![winscale2-mupen](https://user-images.githubusercontent.com/208340/38584899-87c7b3a6-3d20-11e8-9174-a64aee156213.png)

Reicast:

![winscale2-reicast](https://user-images.githubusercontent.com/208340/38584904-902e9744-3d20-11e8-8b3a-020f4e598103.png)

### Version/Commit

- RetroArch: 1.7.1

### Environment information

- OS: FreeBSD 12-CURRENT
- Compiler: clang 6.0.0
- Compositor: Weston [fork](https://github.com/myfreeweb/weston)

**Comment by bparker06** [comment]
paging @Tatsuya79

**Comment by bparker06** [comment]
Actually, he recently made HiDPI changes that are not in 1.7.1, can you try the latest nightly first?

**Comment by Tatsuya79** [comment]
I'm not sure what I changed could have much importance here.
It was mostly about thumbnails scaling and prior to that, some multipliers that react to [xmb_scale_factor which is just that](https://github.com/libretro/RetroArch/blob/master/menu/drivers/xmb.c#L3749).
A multiplier that reacts according to video_info->width.
XMB is probably ignoring the windowed scale setting completely since the beginning.

**Comment by bparker06** [comment]
I think in fullscreen it ignores it but not windowed, not exactly sure what you'd call this though.

**Comment by Tatsuya79** [comment]
It's probably taking the game resolution width divided by 1920 to scale stuff?
I'd need to see how it reacts, I'm always using fullscreen mode and didn't care much about windowed mode until now.

**Comment by myfreeweb** [comment]
Yeah, fullscreen mode with 100 xmb scale is actually usable! A little bit big (but that's fine), and there's a minor bug with multi-line labels:

![fullscreen-xmb](https://user-images.githubusercontent.com/208340/38615918-fcfbe898-3d9a-11e8-940f-ba57cabe08f9.png)

**Comment by Tatsuya79** [comment]
I looked into it but I don't see how to get the real window resolution, xmb is just calling video_driver_get_size.
Scaling of text is problematic too. There's an arbitrary margin in pixels and then the font is displayed there with another scaling value. Everything isn't perfectly relative in the code.

**Comment by i30817** [comment]
I often had to change the mesa driver default fullscreen scaling to avoid 'stretching', but that was on external dosbox.

It was `xrandr --output LVDS1 --set "scaling mode" "Full aspect"` with LVDS1 replaced by the monitor name (output on bare xrandr).

Still, i dunno if xrandr is even relevant to Wayland.

**Comment by myfreeweb** [comment]
I saw this in release notes for 1.7.2:

> (For Linux users) Wayland should now have proper scaling with XMB menu driver.

I think it might actually be worse now? Now in fullscreen mode, everything is large in Vulkan, but small in OpenGL.

**Comment by i30817** [comment]
Now that i'm using Wayland myself i verified my trick is not needed (or supported) there. xrandr still exists, but the property 'scaling mode' doesn't seem to work (but it's not needed since all dosbox games appear to respect aspect ratio in fullscreen).

**Comment by alfrix** [comment]
this should be fixed in the current nightly

**Comment by orbea** [comment]
@alfrix Do you know when it was fixed?

**Comment by alfrix** [comment]
windowed scaling was fixed in 0486979bc9fb87ba72c734e228fead018e3fb72f
and related hotfix 94211bf833e8ec19c7f6a86c92de07813b259b77
